dlt/destinations/dataset.py

```diff
@@ -0,0 +1,96 @@
+from typing import Any, Generator
+
+from contextlib import contextmanager
+from dlt.common.destination.reference import (
+    SupportsReadableRelation,
+    SupportsReadableDataset,
+)
+
+from dlt.common.schema.typing import TTableSchemaColumns
+from dlt.destinations.sql_client import SqlClientBase
+from dlt.common.schema import Schema
+
+
+class ReadableDBAPIRelation(SupportsReadableRelation):
+    def __init__(
+        self,
+        *,
+        client: SqlClientBase[Any],
+        query: str,
+        schema_columns: TTableSchemaColumns = None,
+    ) -> None:
+        """Create a lazy evaluated relation to for the dataset of a destination"""
+        self.client = client
+        self.schema_columns = schema_columns
+        self.query = query
+
+        # wire protocol functions
```

Comment:
I couldn't do much better than this, if you have a nicer implementation of this that still keeps the same interface from the outside, let me know.